File: as/src/androidTest/java/com/trikset/gamepad/KeepAliveTests.java
package com.trikset.gamepad;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import androidx.test.rule.ActivityTestRule;

import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import java.util.Iterator;
import java.util.ListIterator;

@RunWith(JUnit4.class)
public class KeepAliveTests {
    @Rule
    public final ActivityTestRule<MainActivity> mActivityTestRule =
            new ActivityTestRule<>(MainActivity.class);

    @Test
    public void keepAliveShouldBeReceivedAfterGivenTimePeriod() throws InterruptedException {
        final int timeout = 1000;

        DummyServer server = new DummyServer();
        SenderService client = mActivityTestRule.getActivity().getSenderService();

        client.setTarget(DummyServer.IP, DummyServer.DEFAULT_PORT);
        client.setKeepaliveTimeout(timeout);
        // In order to set connection up
        client.send("test");

        Thread.sleep(timeout + 5000);
        server.stopListening();

        ListIterator<String> messages = server.getReceivedMessages().listIterator();
        while (messages.hasNext()) {
            messages.next();
        }
        assertEquals(String.format("keepalive %d", timeout), messages.previous());
    }

    @Test
    public void keepaliveMessagesShouldNotBeSentAfterDisconnect() throws InterruptedException {
        DummyServer server = new DummyServer();
        SenderService client = mActivityTestRule.getActivity().getSenderService();
        client.setTarget(DummyServer.IP, DummyServer.DEFAULT_PORT);
        client.setKeepaliveTimeout(5000);

        // In order to set connection up
        client.send("testtest");
        Thread.sleep(5000);
        client.disconnect("testtest");

        Thread.sleep(1000);
        server.stopListening();

        Iterator<String> messages = server.getReceivedMessages().iterator();
        assertTrue(messages.hasNext());
        assertEquals("testtest", messages.next());
        assertFalse(messages.hasNext());
    }

}

File: app/src/androidTest/kotlin/com/trikset/gamepad2/KeepAliveTests.kt
package com.trikset.gamepad2

import java.util.Locale
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4

@RunWith(JUnit4::class)
class KeepAliveTests {

  // KeepAliveTests never interact with the view hierarchy, so they do not
  // need the focus-wait (and must not be slowed by it on a focus-less CI emulator).
  @get:Rule val mActivityTestRule = FocusAwareActivityTestRule(MainActivity::class.java, false)

  @Test
  @Throws(InterruptedException::class)
  fun keepAliveShouldBeReceivedAfterGivenTimePeriod() {
    val timeout = 1000
    val server = DummyServer()
    val client = mActivityTestRule.activity.senderService
    client.setTarget(DummyServer.IP, DummyServer.DEFAULT_PORT)
    client.keepaliveTimeout = timeout
    // In order to set connection up
    client.send("test")

    // Wait for the keepalive to arrive instead of guessing at a sleep budget.
    assertTrue(server.awaitMessage(String.format(Locale.ROOT, "keepalive %d", timeout), 30000))
    server.stopListening()

    val messages = server.receivedMessages.listIterator()
    while (messages.hasNext()) {
      messages.next()
    }
    assertEquals(String.format(Locale.ROOT, "keepalive %d", timeout), messages.previous())
  }

  @Test
  @Throws(InterruptedException::class)
  fun keepaliveMessagesShouldNotBeSentAfterDisconnect() {
    val server = DummyServer()
    val client = mActivityTestRule.activity.senderService
    client.setTarget(DummyServer.IP, DummyServer.DEFAULT_PORT)
    client.keepaliveTimeout = 2000

    // In order to set connection up
    client.send("testtest")
    assertTrue(server.awaitMessage("testtest", 30000))
    client.disconnect("testtest")

    // Give any (incorrectly scheduled) keepalive a chance to appear within a
    // bounded window, then prove none did (keepalive period is 2000ms).
    assertFalse(server.anyMessageWithin(2500))
    server.stopListening()

    val messages = server.receivedMessages.iterator()
    assertTrue(messages.hasNext())
    assertEquals("testtest", messages.next())
    assertFalse(messages.hasNext())
  }
}
